### PR TITLE
Use `log` and `env_logger`

### LINKED ISSUE #159: Use a logging library
Replace `println!` with [log](https://crates.io/crates/log) and [env_logger](https://crates.io/crates/env_logger)

### PR DESCRIPTION
**Pull request by Schaeff**
Closes #159 

Verbosity can be controlled with the `RUST_LOG` environment variable, see how [here](https://docs.rs/env_logger/latest/env_logger/)

I set everything to debug except the logs which required `verbose` to trace and the progress counter to info. Arbitrary, happy to change.

Note that logs are not captured when running tests (it requires more configuration).

### CODE REVIEW

**Comment by chriseth** [comment]
Hm, I don't really like the fact that you control the logs via environment variables. How does the output look like?

**Comment by Schaeff** [comment]
By default like this

```
RUST_LOG=trace cargo run asm main.asm --force                                
   Compiling powdr v0.1.0 (/Users/thibaut/dev/powdr)
    Finished dev [unoptimized + debuginfo] target(s) in 1.26s
     Running `target/debug/compiler asm main.asm --force`
[2023-04-12T10:01:17Z DEBUG powdr::compiler] Evaluating fixed columns...
[2023-04-12T10:01:17Z DEBUG powdr::compiler] Wrote constants.bin.
[2023-04-12T10:01:17Z DEBUG powdr::compiler] Deducing witness columns...
[2023-04-12T10:01:17Z DEBUG powdr::witness_generator::bit_constraints] Determined the following global bit constraints:
[2023-04-12T10:01:17Z DEBUG powdr::witness_generator::bit_constraints]   Assembly.first_step: 0x1
[2023-04-12T10:01:17Z DEBUG powdr::witness_generator::bit_constraints]   Assembly.line: 0x0
[2023-04-12T10:01:17Z DEBUG powdr::witness_generator::bit_constraints] Determined the following identities to be purely bit/range constraints:
[2023-04-12T10:01:17Z TRACE powdr::witness_generator::generator] ===== Row 0:
        Assembly.x = <unknown>
        Assembly.r = <unknown>
[2023-04-12T10:01:17Z TRACE powdr::witness_generator::generator] ===== Row 0:
        Assembly.x = <unknown>
        Assembly.r = <unknown>
[2023-04-12T10:01:17Z DEBUG powdr::compiler] Wrote commits.bin.
[2023-04-12T10:01:17Z DEBUG powdr::compiler] Wrote ./main.pil.json.
```

Would you rather have it as a CLI argument like now? That's also possible.

**Comment by leonardoalt** [comment]
Yea it looks kinda ugly tbh :grimacing: Can we get rid of the LHS header thing?

**Comment by Schaeff** [comment]
@leonardoalt can you share your dream format and I'll check

**Comment by leonardoalt** [comment]
> @leonardoalt can you share your dream format and I'll check

Like now just text

**Comment by Schaeff** [comment]
Changed to

```
RUST_LOG=trace cargo run asm main.asm --force
   Compiling powdr v0.1.0
    Finished dev [unoptimized + debuginfo] target(s) in 3.07s
     Running `target/debug/compiler asm main.asm --force`
Evaluating fixed columns...
Wrote constants.bin.
Deducing witness columns...
Determined the following global bit constraints:
  Assembly.first_step: 0x1
  Assembly.line: 0x0
Determined the following identities to be purely bit/range constraints:
===== Row 0:
    Assembly.x = <unknown>
    Assembly.r = <unknown>
===== Row 0:
    Assembly.x = <unknown>
    Assembly.r = <unknown>
Wrote commits.bin.
Wrote ./main.pil.json.
```

**Comment by leonardoalt** [comment]
Nice!

**Comment by Schaeff** [comment]
So what granularity do we want and how do we want to control it?

**Review by leonardoalt — APPROVED** [review]

**Comment by leonardoalt** [comment]
> Hm, I don't really like the fact that you control the logs via environment variables. How does the output look like?

I see what you mean, but honestly this is pretty clean. Also we already do this for Rust so, I'd say it's fine.

**Comment by leonardoalt** [comment]
> So what granularity do we want and how do we want to control it?

I'm fine with the current version. I'd say granularity is a number and we can split the code later.

**Comment by chriseth** [comment]
Yeah, we can still tune it. Can you Squash before merge?

**Comment by chriseth** [comment]
Feel free to merge - I would like to adjust the levels a bit but can also do it later.

File: src/bin/compiler.rs [hunk 1]
```diff
@@ -124,71 +106,69 @@ fn split_inputs(inputs: &str) -> Vec<AbstractNumberType> {
 }
 
 fn main() {
+    let mut builder = Builder::from_default_env();
```
[the remark this answers is not in the document]
Comment (Schaeff) [reply src/bin/compiler.rs:109]: Changing to stdout

File: src/bin/compiler.rs [hunk 2]
```diff
@@ -124,71 +106,69 @@ fn split_inputs(inputs: &str) -> Vec<AbstractNumberType> {
...
                 Path::new(&output_directory),
                 force,
-                verbose,
             );
         }
         Commands::Reformat { file } => {
             let contents = fs::read_to_string(&file).unwrap();
             match powdr::parser::parse(Some(&file), &contents) {
-                Ok(ast) => println!("{ast}"),
+                Ok(ast) => log::debug!("{ast}"),
```
[the remark this answers is not in the document]
Comment (Schaeff) [reply src/bin/compiler.rs:160]: Done

File: src/compiler.rs [hunk 3]
```diff
@@ -127,37 +113,35 @@ fn compile(
     file_name: &str,
     output_dir: &Path,
     query_callback: Option<impl FnMut(&str) -> Option<AbstractNumberType>>,
-    verbose: bool,
 ) -> bool {
     let mut success = true;
-    println!("Evaluating fixed columns...");
+    log::debug!("Evaluating fixed columns...");
```
[the remark this answers is not in the document]
Comment (Schaeff) [reply src/compiler.rs:118]: Feel free to change this in your branches

**Pull request merged**